### PR TITLE
make clean should remove .bin and .hex too

### PR DESCRIPTION
**Pull request by fauxpark**
## Description
It didn't seem correct to me that `make clean` leaves the final firmware behind. Also useful when running `make all` for testing stuff.

## Types of changes

- [ ] Core
- [ ] Bugfix
- [ ] New Feature
- [x] Enhancement/Optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/Layout/Userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by this PR

* None.

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document. (https://docs.qmk.fm/#/contributing)
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by vomindoraan** [comment]
It removes the bin and hex files in the `build` directory since it deletes it. I think this is intended, and I find it quite convenient that the built firmware files remain in the base directory when you run `make clean`.

**Comment by fauxpark** [comment]
Ehh, I expect `make clean` to remove *all* build artifacts from the repo directory... this includes the copied .hex/bin files. Otherwise, the state of the directory is not really "clean".

What is the utility of this being the case, then? It just seems to me like it's harder to know whether that particular .hex file is up to date or working or not. And yeah, running make again will simply overwrite it, but still.

**Comment by mechmerlin** [comment]
I've always anticipated `make clean` to remove all build artifacts and the resulting executables including hexes and bins. I think this is a good change but I also agree with Konstantin that it's been very helpful not moving them. lol.

**Comment by pelrun** [comment]
`make clean` generally doesn't bring the repo back to a pristine state, which is why many projects end up with additional `make distclean` or other targets to remove other artifacts as well.

But if you *really* need your working copy to be pristine, you can always use `git clean` to do it...

**Comment by vomindoraan** [comment]
↑ This. The main use case for `make clean` is to force targets/dependencies to recompile if you suspect some of them are bad, or if you've made changes to the makefile. This doesn't necessarily mean completely cleaning the working tree — in QMK's case, it just means cleaning the `.build` directory, which makes sense if you ask me.
You can use `git reset --hard && git clean -dxf` (or `-dxi`) to do a hard clean of the project if you want.

By the way, I just noticed `clean` isn't a `.PHONY` target. This can be a problem if you have an actual file called `clean`. Although this is unlikely, I'd still suggest adding `.PHONY: clean` to the line above `clean:` as it's best practice.
Same applies to the `list-keyboards`, `generate-keyboards-file` and `git-submodule` targets.

File: Makefile [hunk 1]
```diff
@@ -125,9 +125,11 @@ generate-keyboards-file:
 	exit 0
 
 clean:
```
Comment (vomindoraan) [inline Makefile]:
```suggestion
.PHONY: clean
clean:
```
Do this for `.PHONY: list-keyboards` (line 115), `.PHONY: generate-keyboards-file` (line 123) and `.PHONY: git-submodule` (line 582) as well.

**Comment by fauxpark** [comment]
> But if you really need your working copy to be pristine, you can always use `git clean` to do it...

Now I've also removed the new files I've been working on, or the keyboard folders I don't want to PR 👍

I get that `git clean` will do the job, but Git didn't generate that .hex file, make did.

**Comment by jackhumbert** [comment]
Would adding a new `distclean` target work here, instead of modifying the `clean` one?

**Comment by vomindoraan** [comment]
> Now I've also removed the new files I've been working on, or the keyboard folders I don't want to PR 👍

Crap, yeah, we probably should have included a warning. Sorry about that. There are less aggressive versions of `git clean` that won't remove everything outright. For future reference, I recommend using `git clean -i` to do an interactive clean — this will let you pick which files you want to remove and which you want to keep. Add `-d` for directory removal or `-x` for removal of ignored files.
Alternatively, you can use `-n` to do a dry run: this lists which files the clean would remove without actually removing them.


> Would adding a new `distclean` target work here, instead of modifying the `clean` one?

Yes, this would work. The use case also mostly matches the description of `distclean` on [gnu.org](https://www.gnu.org/prep/standards/html_node/Standard-Targets.html):
> `distclean`
> Delete all files in the current directory (or created by this makefile) that are created by configuring or building the program. If you have unpacked the source and built the program without creating any other files, `make distclean` should leave only the files that were in the distribution.

**Comment by fauxpark** [comment]
> Crap, yeah, we probably should have included a warning. Sorry about that.

Oh, I didn't actually do that, don't worry. I was just pointing out that `git clean` gets rid of *all* untracked files in the repo, so while it does achieve the desired result of returning the repo to a pristine state, the idea is to remove all *generated* files.

**Comment by vomindoraan** [comment]
Ah, okay, good! In that case, how about adding a `distclean` target for that like Jack suggested?

File: Makefile [hunk 2]
```diff
@@ -120,14 +121,18 @@ define PRINT_KEYBOARD
...
+.PHONY: clean
 clean:
-	echo -n 'Deleting .build ... '
+	echo 'Deleting .build/ ...'
 	rm -rf $(BUILD_DIR)
-	echo 'done'
+	echo 'Deleting *.bin and *.hex ...'
+	rm -f *.bin *.hex
+	echo 'Done.'
 	exit 0
```
Comment (vomindoraan) [inline Makefile]:
```suggestion

.PHONY: distclean
distclean: clean
	echo -n 'Deleting *.bin and *.hex ... '
	rm -f *.bin *.hex
	echo 'done'

```
Add the `distclean` target.

File: Makefile [hunk 3]
```diff
@@ -112,6 +112,7 @@ $(eval $(call GET_KEYBOARDS))
 # Only consider folders with makefiles, to prevent errors in case there are extra folders
 #KEYBOARDS += $(patsubst $(ROOD_DIR)/keyboards/%/rules.mk,%,$(wildcard $(ROOT_DIR)/keyboards/*/*/rules.mk))
 
+.PHONY: list-keyboards
 list-keyboards:
 	echo $(KEYBOARDS)
 	exit 0
```
Comment (vomindoraan) [inline Makefile]:
```suggestion
```
Remove additional unnecessary `exit`s.

File: Makefile [hunk 4]
```diff
@@ -120,14 +121,18 @@ define PRINT_KEYBOARD
 	$(info $(PRINTING_KEYBOARD))
 endef
 
+.PHONY: generate-keyboards-file
 generate-keyboards-file:
 	$(foreach PRINTING_KEYBOARD,$(KEYBOARDS),$(eval $(call PRINT_KEYBOARD)))
 	exit 0
```
Comment (vomindoraan) [inline Makefile]:
```suggestion
```

File: Makefile [hunk 5]
```diff
@@ -120,14 +121,18 @@ define PRINT_KEYBOARD
...
 
+.PHONY: clean
 clean:
-	echo -n 'Deleting .build ... '
+	echo 'Deleting .build/ ...'
```
[the remark this answers is not in the document]
Comment (vomindoraan) [reply Makefile]: It's how command line tools usually do it. For example, while `git checkout` is in progress:
```
Checking out files:  64% (1071/1673)
```
When it finishes, the same line becomes:
```
Checking out files: 100% (1671/1671), done.
```

**Comment by fauxpark** [comment]
@vomindoraan I note that the GCC module of DMBS, the build system LUFA uses, [has `clean` and `mostlyclean` targets](https://github.com/abcminiuser/dmbs/blob/master/DMBS/gcc.md#build-targets). Admittedly, DMBS doesn't copy the .hex file to the root directory.

**Comment by drashna** [comment]
Thanks!

**Pull request merged**